### PR TITLE
ENH: pad_inches='layout' for savefig

### PR DESCRIPTION
**Pull request by rcomer**
## PR Summary

When using constrained or compressed layout,

```python
savefig(filename, bbox_inches="tight", pad_inches="layout")`
```
will use the padding defined on the layout engine.

Since the layout engine has configurable padding information on it, it makes sense to me that there should be a simple way to tell `savefig` to use that.  This option means that either the height or width of the output file will match what was set at `figsize`.  When preparing figures for a paper, a user might be particularly fussy about the width.  Note that the default pad for `bbox_inches='tight'` is larger than the default for the layout engine, so currently a little fiddling is required to achieve that (I told myself that I shouldn't really care about 3mm but...).

## PR Checklist

**Documentation and Tests**
- [x] Has pytest style unit tests (and `pytest` passes)
- [x] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).
- [ ] New plotting related features are documented with examples.

**Release Notes**
- [ ] New features are marked with a `.. versionadded::` directive in the docstring and documented in `doc/users/next_whats_new/`
- [N/A] API changes are marked with a `.. versionchanged::` directive in the docstring and documented in `doc/api/next_api_changes/`
- [ ] Release notes conform with instructions in  `next_whats_new/README.rst` or `next_api_changes/README.rst`

### CODE REVIEW

File: lib/matplotlib/transforms.py [hunk 1]
```diff
@@ -616,10 +616,22 @@ def expanded(self, sw, sh):
         a = np.array([[-deltaw, -deltah], [deltaw, deltah]])
         return Bbox(self._points + a)
 
-    def padded(self, p):
-        """Construct a `Bbox` by padding this one on all four sides by *p*."""
+    def padded(self, pw, ph=None):
```
Comment (oscargus) [inline lib/matplotlib/transforms.py:619]: After just expiring a number of deprecations which involved name changes of non-optional arguments, I would suggest adding a `@_api.rename_parameter("3.8", "p", "pw")`.

In addition, may one should go all in and spell out the parameter name?

File: lib/matplotlib/transforms.py [hunk 2]
```diff
@@ -616,10 +616,22 @@ def expanded(self, sw, sh):
...
+        """
+        Construct a `Bbox` by padding this one on all four sides.
+
+        Parameters
+        ----------
+        pw : float
+            Width pad
+        ph: float, optional
+            Height pad.  Defaults to pw.
```
Comment (oscargus) [inline lib/matplotlib/transforms.py:628]:
```suggestion
            Height pad.  Defaults to *pw*.
```

File: lib/matplotlib/transforms.py (hunk 1 above)
Comment (rcomer) [reply lib/matplotlib/transforms.py:619]: I picked very short names for consistency with (some of) the other methods on this class, but maybe `w_pad` and `h_pad` would be better?  That would make them consistent with those for the relatively recently added constrained layout engine.

Thanks for the pointer on `_api.rename_parameter`.  I tend to forget that the difference between positional and keyword args in python is only pretend!

**Comment by jklymak** [comment]
I'm not following this feature, but didn't look too closely.  Your example calls compressed layout on a single axes (??) and then trims the saved figure by a pad.  However, I'm not following why you don't just do

```python
fig, axs = plt.subplots(2, 2, layout='compressed')
for ax in axs.flat:
    ax.set_aspect(1)

fig.savefig('boo.png', bbox_inches='tight', pad_inches=0.05)
```

I also didn't understand the motivation vis-a-vis a certain width figure?  If the user wants a compressed layout and a figure of a certain size they need only do `fig, axs = plt.subplot(2, 2, layout='compressed', figsize=(4.2367, 20))` where 20 is big enough that the compression always happens in the vertical.  

Overall, I'd be very leery of a feature that changes the visual appearance of a figure for just `savefig`.   With a name like `bbox_inches='compressed'` I'd _expect_ compressed layout to be used on saving, whether or not it was used previously.  I don't see that you are doing that (?), but...

If the motivation is to have different vertical and horizontal padding, why not make that an option to `pad_inches`?

**Comment by rcomer** [comment]
My test is not very representative of my target use-case because I was just going for "what's the simplest test I can make that verifies the outer padding looks how I expect".  I set the `h_pad` because I wanted to check what happens if `h_pad` and `w_pad` are different.

In my real use-case I am more than happy with the layout engine's default pad, so currently I can get what I want with something like

```python
fig, axs = plt.subplots(2, 2, figsize=(my_chosen_width, some_large_height), layout='compressed')
for ax in axs.flat:
    ax.set_aspect(1)

fig.savefig('boo.png', bbox_inches='tight',
            pad_inches=plt.rcParams["figure.constrained_layout.h_pad"])
```

which was all easy to figure out _except_ what exactly to use for `pad_inches`.  In general the layout engine Just Works in the background so I'd never even needed to learn what class of object does the work before.  So although it was obvious that compressed layout must have a pad stored somewhere, it took me some searching in the docs to figure out where.  Even once you know, passing in the rcParam feels a little cumbersome (and if I just pass `0.04167` I will definitely need to look it up every time!)  So I think it would be nice to provide some sort of shortcut that says "the padding on save should match the padding on the layout engine".

I picked `bbox_inches="compressed"` because I was thinking it only makes sense to use this with compressed layout.  Thought process was essentially "this is needed when you have fixed aspect ratios, and if you have fixed aspect ratios you'll be using compressed layout".  But thinking again that's clearly not true.  So yes, this would need some more thought on the applicability and therefore on the name.

Thinking again this morning, I think my ideal solution would be to change the `pad_inches` default.  So if you are using a `ConstrainedLayoutEngine` it automatically gets the pad off that, and if not it falls back to the current default of `rcParams["savefig.pad_inches"]`.

**Comment by jklymak** [comment]
> Thinking again this morning, I think my ideal solution would be to change the `pad_inches` default. So if you are using a `ConstrainedLayoutEngine` it automatically gets the pad off that, and if not it falls back to the current default of `rcParams["savefig.pad_inches"]`.

I think modifying `pad_inches` is a good path to take.  Whether the default should change is another issue, but something like `pad_inches='layout'` might be good for using the layout engine defaults.

**Comment by rcomer** [comment]
OK I've now implemented `pad_inches="layout"`.  Currently it only works if a `ConstrainedLayoutEngine` is in use.  If `pad_inches="layout"` is passed when `ConstrainedLayoutEngine` is not in use it silently falls back to using the default pad.  I'm not sure whether it's better to do that or raise an exception.  It could also be implemented for `TightLayoutEngine`, but I do not know how useful that would be: I would not envisage using it with tight layout myself.

File: lib/matplotlib/backend_bases.py [hunk 3]
```diff
@@ -2341,9 +2344,15 @@ def print_figure(
                 if bbox_inches == "tight":
                     bbox_inches = self.figure.get_tightbbox(
                         renderer, bbox_extra_artists=bbox_extra_artists)
-                    if pad_inches is None:
-                        pad_inches = rcParams['savefig.pad_inches']
-                    bbox_inches = bbox_inches.padded(pad_inches)
+                    if (isinstance(layout_engine, ConstrainedLayoutEngine) and
+                            pad_inches == "layout"):
+                        h_pad = layout_engine.get()["h_pad"]
+                        w_pad = layout_engine.get()["w_pad"]
```
Comment (jklymak) [inline lib/matplotlib/backend_bases.py:2350]: the layout_engine pads are in figure-normalized units (or at least that is what the docs say, if that isn't the case, then we should fix the docs), whereas bbox_inches='tight' pads are supposed to be in inches...

**Review by jklymak — COMMENTED** [review]
I don't think this is quite right, though I didn't upload to debug yet.

File: lib/matplotlib/backend_bases.py (hunk 3 above)
Comment (jklymak) [reply lib/matplotlib/backend_bases.py:2350]: Oooops, thats not good.  I guess padding in inches (or points) makes more sense to me (ahem, and probably did when I wrote this) and the docs need to be fixed.

Comment (rcomer) [reply lib/matplotlib/backend_bases.py:2350]: The matplotlibrc is OK though
```
#figure.constrained_layout.h_pad:  0.04167  # Padding around axes objects. Float representing
#figure.constrained_layout.w_pad:  0.04167  # inches. Default is 3/72 inches (3 points)
```

**Review by jklymak — APPROVED** [review]

**Review by oscargus — APPROVED** [review]

**Comment by oscargus** [comment]
Maybe a bit too early to approve as I think there should go some `.. versionadded::` and maybe a user release note in. But other than that, looks good!

**Comment by jklymak** [comment]
Should this get a what's new entry?

**Comment by QuLogic** [comment]
No, the branch is made; anything for 3.8 is fine to put in `main`.

**Comment by jklymak** [comment]
I wouldn't add a version added note, because it's just a new argument to the keyword.  But I'd add to the what's-new.

**Comment by rcomer** [comment]
What's new entry now added here:
https://output.circle-artifacts.com/output/job/71c306bd-1f01-4f4f-bbb8-8c4b9a6942fc/artifacts/0/doc/build/html/users/next_whats_new/savefig_bbox_layout.html

**Comment by rcomer** [comment]
Thanks @jklymak, @oscargus

**Pull request merged**